README.md
# Announcement

We are rising funds for next version (2.0) of ESP-RFID.

<p align="center">
  <img src="https://opencollective.com/esp-rfid/tiers/esp-rfid-user.svg?avatarHeight=56">
</p>

What to expect from version 2.0:

* Better Wi-Fi connection handling
* MQTT Subscribing (activate relay on MQTT event)
* Better handling of hardware state (like health checks)
* Time Based One Time Password access (keypad abstraction layer)
* More polished WebUI
* Yet still Open Source (MIT License)
* Optional Goal --> NFC NDEF transactions from mobile phone to esp-rfid and vice versa

---


# ESP RFID - Access Control with ESP8266, RC522 PN532 Wiegand RDM6300

[![Backers on Open Collective](https://opencollective.com/esp-rfid/backers/badge.svg)](#backers) [![Sponsors on Open Collective](https://opencollective.com/esp-rfid/sponsors/badge.svg)](#sponsors) [![Chat at https://gitter.im/esp-rfid/Lobby](https://badges.gitter.im/esp-rfid.svg)](https://gitter.im/esp-rfid/Lobby) [![Build Status](https://travis-ci.org/esprfid/esp-rfid.svg?branch=stable)](https://travis-ci.org/esprfid/esp-rfid) [![Codacy Badge](https://api.codacy.com/project/badge/Grade/d475052a09bd4865bc2654f524614cba)](https://www.codacy.com/app/omersiar/esp-rfid?utm_source=github.com&amp;utm_medium=referral&amp;utm_content=esprfid/esp-rfid&amp;utm_campaign=Badge_Grade) [![BCH compliance](https://bettercodehub.com/edge/badge/esprfid/esp-rfid?branch=stable)](https://bettercodehub.com/) [![Bountysource](https://api.bountysource.com/badge/team?team_id=242217)](https://salt.bountysource.com/checkout/amount?team=esp-rfid)

Access Control system using a cheap MFRC522, PN532 RFID, RDM6300 readers or Wiegand RFID readers and Espressif's ESP8266 Microcontroller.

[See Demo Here](https://bitadvise.com/esp-rfid/)

[![Showcase Gif](https://raw.githubusercontent.com/esprfid/esp-rfid/stable/demo/showcase.gif)](https://bitadvise.com/esp-rfid/)[![Board](https://raw.githubusercontent.com/esprfid/esp-rfid/stable/demo/board.jpg)](https://www.tindie.com/products/nardev/esp-rfid-relay-board-12v-in-esp8266-board/)

## Features
### For Users
* Minimal effort for setting up your Access Control system, just flash and everything can be configured via Web UI
* Capable of managing up to 1.000 Users (even more is possible)
* Great for Maker Spaces, Labs, Schools, etc
* Cheap to build and easy to maintain
### For Tinkerers
* Open Source (minimum amount of hardcoded variable, this means more freedom)
* Using WebSocket protocol to exchange data between Hardware and Web Browser
* Data is encoded as JSON object
* Records are Timestamped (Time synced from a NTP Server)
* MQTT enabled
* Bootstrap, jQuery, FooTables for beautiful Web Pages for both Mobile and Desktop Screens
* Thanks to ESPAsyncWebServer Library communication is Asynchronous
### Official Hardware
* Small size form factor, sometimes it is possible to glue it into existing readers.
* Single power source to power 12V/2A powers ESP12 module, RFID Wiegand Reader and magnetic lock for opening doors.
* Exposed programming pins for ESP8266
* Regarding hardware design, you get multiple possible setup options:
* Forward Bell ringing on reader to MCU or pass it out of board
* Track Door Status
* Control reader’s status LED
* Control reader’s status BUZZER sound *
* Power reader, lock and the board through single 12V, 2A PSU
* Optionally power magnetic lock through external AC/DC PSU
* Possible to use any kind and any type of Wiegand readers
* Enables you to make IOT Access System with very little wiring
* Fits in an universal enclosures with DIN mount
* Open Source Hardware

Get more information and see accessory options from [Tindie Store](https://www.tindie.com/products/nardev/esp-rfid-relay-board-12v-in-esp8266-board/)

| What are others saying about esp-rfid? |
| ---- |
| _“Hi, nice project.”_ – [@Rotzbua]() |
| _“Your app works like a charm”_ – [@tueddy ]() |
| _“Just stumbled upon this project while planning to do something similar. Very beautifully done!”_ – [@LifeP]() |
| _“Hello, I've come across your project and first of all… wow - thanks to all contributors for your hard work!”_ – [@byt3w4rri0r]() |
| _“Brilliant work.”_ – [@danbicks]() |
| _“This is an impressive project.”_ – [@appi1]() |
| _“I'd like to thank every single contributor for creating this epic project.”_ – [@TheCellMc]() |
| _“Congratulations for your awesome work! This project is absolutely brilliant.”_ – [@quikote]() |

## Getting Started
This project still in its development phase. New features (and also bugs) are introduced often and some functions may become deprecated. Please feel free to comment or give feedback.

* [See Demo Here](https://bitadvise.com/esp-rfid/)
* Get the latest release from [here.](https://github.com/esprfid/esp-rfid/releases)
* See [Known Issues](https://github.com/esprfid/esp-rfid#known-issues) before starting right away.
* See [Security](https://github.com/esprfid/esp-rfid#security) for your safety.
* See [ChangeLog](https://github.com/esprfid/esp-rfid/blob/dev/CHANGELOG.md)

### What You Will Need
### Hardware
* [Official ESP-RFID Relay Board](https://www.tindie.com/products/nardev/esp-rfid-relay-board-12v-in-esp8266-board/)
or
* An ESP8266 module or a development board like **WeMos D1 mini** or **NodeMcu 1.0** with at least **32Mbit Flash (equals to 4MBytes)** (ESP32 does not supported for now)
* A MFRC522 RFID PCD Module or PN532 NFC Reader Module or RDM6300 125KHz RFID Module Wiegand based RFID reader
* A Relay Module (or you can build your own circuit)
* n quantity of Mifare Classic 1KB (recommended due to available code base) PICCs (RFID Tags) equivalent to User Number

### Software

#### Using Compiled Binaries
Download compiled binaries from GitHub Releases page
https://github.com/esprfid/esp-rfid/releases
On Windows you can use **"flash.bat"**, it will ask you which COM port that ESP is connected and then flashes it. You can use any flashing tool and do the flashing manually. The flashing process itself has been described at numerous places on Internet.

#### Building With PlatformIO
##### Backend
The build environment is based on [PlatformIO](http://platformio.org). Follow the instructions found here: http://platformio.org/#!/get-started for installing it but skip the ```platform init``` step as this has already been done, modified and it is included in this repository. In summary:

```
sudo pip install -U pip setuptools
sudo pip install -U platformio
git clone https://github.com/esprfid/esp-rfid.git
cd esp-rfid
platformio run
```

When you run ```platformio run``` for the first time, it will download the toolchains and all necessary libraries automatically.

##### Useful commands:

* ```platformio run``` - process/build all targets
* ```platformio run -e generic -t upload``` - process/build and flash just the ESP12e target (the NodeMcu v2)
* ```platformio run -t clean``` - clean project (remove compiled files)

The resulting (built) image(s) can be found in the directory ```/bin``` created during the build process.

##### Frontend
You can not simply edit Web UI files because you will need to convert them to C arrays, which can be done automatically by a gulp script that can be found in tools directory or you can use compiled executables at the same directory as well (for Windows PCs only).

If you want to edit esp-rfid's Web UI you will need (unless using compiled executables):
* NodeJS
* npm (comes with NodeJS installer)
* Gulp (can be installed with npm)

Gulp script also minifies HTML and JS files and compresses (gzip) them.

To minify and compress the frontend, enter the folder ```tools/webfilesbuilder``` and:
* Run ```npm install``` to install dependencies
* Run ```npm start``` to compress the web UI to make it ready for the ESP

In order to test your changes without flashing the firmware you can launch websocket emulator which is included in tools directory.
* You will need to Node JS for websocket emulator.
* Run ```npm install``` to install dependencies
* Run emulator  ```node wserver.js```

There are two alternative ways to test the UI
1. you can launch your browser with CORS disabled:
  ```chrome.exe --args --disable-web-security -–allow-file-access-from-files --user-data-dir="C:\Users\USERNAME"```
  and then open the HTML files directly (Get more information [here](https://stackoverflow.com/questions/3102819/disable-same-origin-policy-in-chrome))
2. alternatively, you can launch a web server from the ```src/websrc``` folder, for example with Python, like this:
  ```python3 -m http.server```
  and then visit ```http://0.0.0.0:8000/```

When testing locally, use the password ```neo``` for admin capabilities.


### Pin Layout

The following table shows the typical pin layout used for connecting readers hardware to ESP:

| ESP8266 | NodeMcu/WeMos | Wiegand | PN532 | MFRC522 | RDM6300 |
|--------:|:-------------:|:-------:|:-----:|:-------:|:-------:|
| GPIO-15 | D8            |         | SS    | SDA/SS  |         |
| GPIO-13 | D7            | D0      | MOSI  | MOSI    |         |
| GPIO-12 | D6            | D1      | MISO  | MISO    |         |
| GPIO-14 | D5            |         | SCK   | SCK     |         |
| GPIO-04 | D2            |         |       |         |         |
| GPIO-05 | D1            |         |       |         |         |
| GPIO-03 | RX            |         |       |         | TX      |

For Wiegand based readers, you can configure D0 and D1 pins via settings page. By default, D0 is GPIO-4 and D1 is GPIO-5

### Steps
* First, flash firmware (you can use /bin/flash.bat on Windows) to your ESP either using Arduino IDE or with your favourite flash tool
* (optional) Fire up your serial monitor to get informed
* Search for Wireless Network "esp-rfid-xxxxxx" and connect to it (It should be an open network and does not require password)
* Open your browser and type either "http://192.168.4.1" or "http://esp-rfid.local" (.local needs Bonjour installed on your computer) on address bar.
* Log on to ESP, default password is "admin"
* Go to "Settings" page
* Configure your amazing access control device. Push "Scan" button to join your wireless network, configure RFID hardware, Relay Module.
* Save settings, when rebooted your ESP will try to join your wireless network.
* Check your new IP address from serial monitor and connect to your ESP again. (You can also connect to "http://esp-rfid.local")
* Go to "Users" page
* Scan a PICC (RFID Tag) then it should glimpse on your Browser's screen.
* Type "User Name" or "Label" for the PICC you scanned.
* Choose "Allow Access" if you want to
* Click "Add"
* Congratulations, everything went well, if you encounter any issue feel free to ask help on GitHub.

### MQTT
You can integrate ESP-RFID with other systems using MQTT. Read the [additional documentation](./README-MQTT.md) for all the details.

### Known Issues
* You need to connect your MFRC522 reader to your ESP properly or you will end up with a boot loop
* Please also check [GitHub issues](https://github.com/esprfid/esp-rfid/issues).

#### Time
We are syncing time from a NTP Server (in Client -aka infrastructure- Mode). This will require ESP to have an Internet connection. Additionally your ESP can also work without Internet connection too (Access Point -aka Ad-Hoc- Mode),  without giving up functionality.
This will require you to do syncing manually. ESP can store and hold time for you approximately 51 days without a major issue, device time can drift from actual time depending on usage, temperature, etc.
So you have to login to settings page and sync it in a timely fashion.

## **Security**
We assume **ESP-RFID** project -as a whole- does not offer strong security. There are PICCs available that their UID (Unique Identification Numbers) can be set manually (Currently esp-rfid relies only UID to identify its users). Also there may be a bug in the code that may result free access to your belongings. And also, like every other network connected device esp-rfid is vulnerable to many attacks including Man-in-the-middle, Brute-force, etc.

This is a simple, hobby grade project, do not use it where strong security is needed.

What can be done to increase security? (by you and by us)

* We are working on more secure ways to Authenticate RFID Tags.
* You can disable wireless network to reduce attack surface. (This can be configured in Web UI Settings page)
* Choose a strong password for the Web UI

## Scalability
Since we are limited on both flash and ram size things may get ugly at some point in the future. You can find out some test results below.

### Tests

#### 1) How many RFID Tag can be handled?
Restore some randomly generated user data on File System worth:

* 1000 separate "userfile"
* random 4 Bytes long UID and
* random User Names and
* 4 bytes random Unix Time Stamp
* each have "access type" 1 byte integer "1" or "0".

Total 122,880 Bytes

At least 1000 unique User (RFID Tag) can be handled, the test were performed on WeMos D1 mini.

#### Additional testing is needed:

* Logging needs testing. How long should it need to log access? What if a Boss needs whole year log?
* Reliability on Flash (these NOR Flash have limited write cycle on their cells). It depends on manufacturer choice of Flash Chip and usage.

## Community
* [![Chat at https://gitter.im/esp-rfid/Lobby](https://badges.gitter.im/esp-rfid.svg)](https://gitter.im/esp-rfid/Lobby) Join community chat on Gitter

### Projects that is based on esp-rfid
* [ESP-IO](https://github.com/Pako2/EventGhostPlugins/tree/master/ESP-IO) Project to manipulate GPIOs with EventGhost
* [ESP-RCM](https://github.com/Pako2/esp-rcm) Room Climate Monitor with ESP8266, HTU21D, Si7021, AM2320

[ESP-RFID-PY](https://github.com/esprfid/esp-rfid-py) Micro-Python implementation of esp-rfid is also made available by @iBobik

### Acknowledgements

- @rneurink
- @thunderace
- @zeraien
- @nardev
- @romanzava
- @arduino12
- @Pako2
- @marelab

See [ChangeLog](https://github.com/esprfid/esp-rfid/blob/dev/CHANGELOG.md)

## Donations
[![OC](https://opencollective.com/esp-rfid/tiers/esp-rfid-user.svg?avatarHeight=56)](https://opencollective.com/esp-rfid)

Developing fully open, extensively tested embedded software is hard and time consuming work. Please consider making donations to support developers behind this beautiful software.

Donations **transparently** processed by **[Open Collective](https://opencollective.com/how-it-works)** and expenses are being made public by OC's open ledger.

* 2017-10-03 [steinar-t](https://github.com/steinar-t)
* 2017-12-10 [saschaludwig](https://github.com/saschaludwig)
* 2018-10-02 Dennis Parsch
* 2019-01-12 Chris-topher Slater
* 2019-04-23 Klaus Blum
* 2019-04-25 Andre Dieteich

## Contributors

This project exists thanks to all the people who contribute.
<a href="https://github.com/esprfid/esp-rfid/graphs/contributors"><img src="https://opencollective.com/esp-rfid/contributors.svg?width=890&button=false" /></a>


## Backers

Thank you to all our backers! 🙏 [[Become a backer](https://opencollective.com/esp-rfid#backer)]

<a href="https://opencollective.com/esp-rfid#backers" target="_blank"><img src="https://opencollective.com/esp-rfid/backers.svg?width=890"></a>


## Sponsors

Support this project by becoming a sponsor. Your logo will show up here with a link to your website. [[Become a sponsor](https://opencollective.com/esp-rfid#sponsor)]

<a href="https://opencollective.com/esp-rfid/sponsor/0/website" target="_blank"><img src="https://opencollective.com/esp-rfid/sponsor/0/avatar.svg"></a>
<a href="https://opencollective.com/esp-rfid/sponsor/1/website" target="_blank"><img src="https://opencollective.com/esp-rfid/sponsor/1/avatar.svg"></a>

CHANGELOG.md
# Change Log
All notable changes to this project will be documented in this file.

## Unreleased
#### Added
- [firmware] Initial support for Wiegand Keypads @frenchie71
- [firmware] Initial support for OTP (TOTP) @frenchie71
- [webui] Added Download function for log files @frenchie71
- [webui] Log file options on WebUI @frenchie71
- [firmware] Log Maintenance Split, Rollover, View, Delete @frenchie7

#### Changed
- [firmware] Wi-Fi connection routine is now improved @frenchie71
- [firmware] Log file operations are now more robust @frenchie71

## [1.3.3] Test-Release 2019-06-22
#### Fixed
* PlatformIO **ISR not in IRAM!** Hotfix

## [1.3.1] Test-Release 2019-06-21
#### Added
- [firmware] Support for Door Status tracking with log and mqtt options @nardev
- [firmware] adding mqtt_publish_info, generateUid functions @nardev
- [firmware] MQTT functionality is greatly improved thanks to @marelab

Reading all user data over MQTT
Sending User data to RFID-DOOR/ESP-RFID over MQTT
Sending door open command over MQTT
Sending door status over MQTT as event
Sending Sync of a RFID-DOOR (IP/Hostname) over MQTT
Configure Sync interval over ESP-RFID GUI
Deleting all User of a ESP-RFID device over MQTT
Sending log event & access data over MQTT

#### Changed
- [firmware] Renaming some variable names @nardev
- [firmware] order of settings in web and hr lines @nardev

#### Fixed
- [webui] Make possible to connect WebUI behind NAT-ed network @Pako2 @abdulhanananwari
- [firmware] Issue #240 "Backup User Data only returns some users" @nardev

## [1.0.2] 2019-01-28

### BREAKING CHANGES (These changes will break your data on device, please make sure made a backup, also you can not use your old settings on this release but only can restore user data)

#### Added
- [firmware] Open/Close Button support @donatmarko
- [firmware] Logging Open/Close Button @donatmarko
- [firmware] the VERSION string is now defined at the beginning of main.cpp @Pako2
- [firmware] #218 Added latching relay support @donatmarko
- [firmware] #189 Flash layout changed to 2MB Firmware / 2MB SPIFFS Data @Pako2
- [firmware] Support for RDM6300 RFID readers (125kHz, UART) #163 @arduino12 / concurrently by @Pako2
- [firmware] debug firmware for debugging purposes
- [tools] executables for tools (no longer need to have node js and gulp for web ui development - **only lightly tested**)
- [firmware] LED_BUILTIN lights up while wifi connected and flashes when it waits for wifi @Pako2
- [webui] IP address choice option in AP mode @Pako2
- [webui] favicon.ico @Pako2
- [tools] websocket emulator can now store new configuration temporarly
- [firmware] log for firmware update #152
- [webui] Expired access attempts logged as "Expired"

#### Fixed
- [firmware] UART Monitor speed @donatmarko
- [firmware] avoid double Serial.begin @Pako2
- [firmware] removing redundant terminating null character  @Pako2
- [firmware] fix the loadconfiguration loop @Pako2
- [firmware] not able to connect MQTT server #157 @fivosg
- [firmware] a MQTT message typo #157 @wamboin23
- [webui] some breaks on web pages
- [webui] usage of !important CSS rule
- [firmware] #191 relay type inversion @Pako2
- [firmware] #190 Increase PN532::WaitReady debug level @Pako2
- [firmware] validuntil is being ignored #151
- [firmware] the boot loop when ssid is empty on configuration file (actually more a workaround than a fix) #154

#### Changed
- [firmware] unify output format debug print of PICC @Pako2
- [firmware] Improve onWsEvent() function @Pako2
- [build] Change release type to a zip file (was tar.gz before)
- [webui] scrollbar on desktop screens (now hidden)
- [webui] sidebar colors (i hope you like it, standart bootstrap color)
- [webui] sanity check for firmware update file #152
- [firmware] lock MFRC522 library to version 1.4.1

## [0.8.1] 2018-09-01
#### Added
- [firmware] Global websocket message to inform ws clients to toggle relay (upcoming client version will use it)
- [webui] incremental id for event log

#### Fixed
- [webui] available flash space calculation
- [webui] #143 duplicate records on User Data backup
- [firmware] #140 MQTT Heartbeat
- [webui] official board's hardware settings did not populate

#### Changed
- [firmware] do not initialize serial output unless we are debugging
- [build] slice main.cpp to multiple parts for better readability
- [webui] Access Type Active to Always
- [firmware] more reliable activation of relay

## [0.8.0]
#### Breaking Changes
- [firmware] Flash partition is changed to 1+3 !!! You need to backup your settings and users before updating to this version
- [firmware] For wiegand readers card id's changed hexadecimal to decimal !!! You need to change hexadecimal values to decimal values on your user backup file

#### Added
- [build] Optimize code for official board
- [firmware] mqtt boot, hearthbeat, access message added

#### Fixed
- [firmware] #128 Do not retain MQTT publishes
- [firmware] Compiller warnings fixed

#### Changed
- [webui] Default wifi type to AP
- [firmware] MQTT Messages are now plain JSON encoded texts
- [build] flash.bat file now asks which firmware to flash

## [0.7.6] - 2018-07-13
#### Fixed
- [firmware] #98 WDT Reset

## [0.7.5] - 2018-06-09
#### Fixed
- [build] BearSSL dependency error with Platformio
- [webui] #115 version numbering
- [firmware] #101 Permanent AP Mode

#### Changed
- [build] Updated PlatformIO configuration file for next PIO release

## [0.7.4] - 2018-04-28
#### Fixed
- [webui] comment out access types
- [firmware] fix a compile warning
- [firmware] change startAP behaviour
- [firmware] publish username to MQTT broker
- [firmware] mqtt username password memory collusion

#### Removed
- [firmware] Modified header

## [0.7.3] - 2018-04-04
#### Added
- [firmware] More debug messages

#### Fixed
- [webui] minor fixes suggested by Codacy
- [webui] get javascript values as real integers

## [0.7.2] - 2018-04-02
#### Fixed
- [webui] MQTT is enabled by default.

## [0.7.1] - 2018-03-30
#### Added

tools/wsemulator/wserver.js
console.log("[ INFO ] Starting ESP-RFID WebSocket Emulation Server");

const WebSocket = require("ws");

console.log("[ INFO ] You can connect to ws://localhost (default port is 8080)");

const wss = new WebSocket.Server({
    port: 8080
});

wss.broadcast = function broadcast(data) {
    wss.clients.forEach(function each(client) {
        if (client.readyState === WebSocket.OPEN) {
            client.send(JSON.stringify(data));
        }
    });
};

var networks = {
    "command": "ssidlist",
    "list": [{
            "ssid": "Company's Network",
            "bssid": "4c:f4:39:a1:41",
            "rssi": "-84"
        },
        {
            "ssid": "Home Router",
            "bssid": "8a:e6:63:a8:15",
            "rssi": "-42"
        },
        {
            "ssid": "SSID Shown Here",
            "bssid": "8a:f5:86:c3:12",
            "rssi": "-77"
        },
        {
            "ssid": "Great Wall of WPA",
            "bssid": "9c:f1:90:c5:15",
            "rssi": "-80"
        },
        {
            "ssid": "Not Internet",
            "bssid": "8c:e4:57:c5:16",
            "rssi": "-87"
        }
    ]
}

var latestlog = {
    "command": "latestlist",
    "page": 1,
    "haspages": 1,
    "list": [
        "{\"timestamp\":1518198383,\"uid\":\"8ab424c10\",\"username\":\"Moody Bond\",\"acctype\":1}",
        "{\"timestamp\":1514952461,\"uid\":\"4de212c96\",\"username\":\"Unknown\",\"acctype\":0}",
        "{\"timestamp\":1516598710,\"uid\":\"8de284c27\",\"username\":\"Marta Cooley\",\"acctype\":99}",
        "{\"timestamp\":1516649998,\"uid\":\"4db504c86\",\"username\":\"Simmons Sosa\",\"acctype\":1}",
        "{\"timestamp\":1517133201,\"uid\":\"9db178a36\",\"username\":\"Jimmie Sheppard\",\"acctype\":1}",
        "{\"timestamp\":1516257556,\"uid\":\"4cf690a75\",\"username\":\"Rutledge Murray\",\"acctype\":1}",
        "{\"timestamp\":1515661586,\"uid\":\"4ab792d39\",\"username\":\"Unknown\",\"acctype\":0}",
        "{\"timestamp\":1515524537,\"uid\":\"9cf869a85\",\"username\":\"Rollins Villarreal\",\"acctype\":1}",
        "{\"timestamp\": 1515823122,\"uid\":\"9db221c40\",\"username\": \"Hayden Baird\",\"acctype\": 1}",
        "{\"timestamp\": 1515066066,\"uid\": \"8cb891d34\",\"username\": \"Tucker Boyer\",\"acctype\": 1}"
    ]
}

var eventlog = {
    "command": "eventlist",
    "page": 1,
    "haspages": 1,
    "list": [
        "{ \"type\": \"WARN\", \"src\": \"sys\", \"desc\": \"Event log cleared!\", \"data\": \"\", \"time\": 1520523010 }",
        "{ \"type\": \"WARN\", \"src\": \"sys\", \"desc\": \"Event log cleared!\", \"data\": \"\", \"time\": 1520523010 }",
        "{ \"type\": \"INFO\", \"src\": \"wifi\", \"desc\": \"WiFi is connected\", \"data\": \"SMC\", \"time\": 13 }",
        "{ \"type\": \"INFO\", \"src\": \"sys\", \"desc\": \"System setup completed, running\", \"data\": \"\", \"time\": 13 }",
        "{ \"type\": \"INFO\", \"src\": \"wifi\", \"desc\": \"WiFi is connected\", \"data\": \"SMC\", \"time\": 13 }",
        "{ \"type\": \"INFO\", \"src\": \"sys\", \"desc\": \"System setup completed, running\", \"data\": \"\", \"time\": 13 }",
        "{ \"type\": \"WARN\", \"src\": \"websrv\", \"desc\": \"New login attempt\", \"data\": \"\", \"time\": 1520578744 }",
        "{ \"type\": \"INFO\", \"src\": \"websrv\", \"desc\": \"Login success!\", \"data\": \"\", \"time\": 1520578744 }",
        "{ \"type\": \"INFO\", \"src\": \"wifi\", \"desc\": \"WiFi is connected\", \"data\": \"SMC\", \"time\": 13 }",
        "{ \"type\": \"INFO\", \"src\": \"sys\", \"desc\": \"System setup completed, running\", \"data\": \"\", \"time\": 13 }",
        "{ \"type\": \"WARN\", \"src\": \"websrv\", \"desc\": \"New login attempt\", \"data\": \"\", \"time\": 1520583560 }"
    ]
}

var users = [{
        "uid": "12c9298d",
        "pincode": "1234",
        "username": "Nikki Kidd",
        "acctype": 0,
        "validuntil": 4562316475
    },
    {
        "uid": "70f3675d",
        "pincode": "1234",
        "username": "Maria Cohen",
        "acctype": 1,
        "validuntil": 1048061447
    },
    {
        "uid": "80c2489b",
        "pincode": "1234",
        "username": "Hilda Whitaker",
        "acctype": 1,
        "validuntil": 9154328938
    },
    {
        "uid": "14a8590b",
        "pincode": "1234",
        "username": "Eloise Floyd",
        "acctype": 1,
        "validuntil": 7903845677
    },
    {
        "uid": "84f2342d",
        "pincode": "1234",
        "username": "Grimes Davenport",
        "acctype": 0,
        "validuntil": 4362181899
    },
    {
        "uid": "53f5031b",
        "pincode": "1234",
        "username": "Wanda Jones",
        "acctype": 1,
        "validuntil": 8077084354
    },
    {
        "uid": "72c4667c",
        "pincode": "1234",
        "username": "Weaver Sampson",
        "acctype": 0,
        "validuntil": 9712710211
    },
    {
        "uid": "23a9198d",
        "pincode": "1234",
        "username": "Nash Hernandez",
        "acctype": 1,
        "validuntil": 8752373818
    },
    {
        "uid": "24c8875b",
        "pincode": "1234",
        "username": "Elvia Vance",
        "acctype": 1,
        "validuntil": 1956504821
    },
    {
        "uid": "91c3021b",
        "pincode": "1234",
        "username": "Lucinda Fischer",
        "acctype": 1,
        "validuntil": 3697215988
    },
    {
        "uid": "19c2258d",
        "pincode": "1234",
        "username": "Gregory Cline",
        "acctype": 1,
        "validuntil": 4515022519
    },
    {
        "uid": "65f1947b",
        "pincode": "1234",
        "username": "Lidia Hancock",
        "acctype": 1,
        "validuntil": 2333382745
    },
    {
        "uid": "21c9851d",
        "pincode": "1234",
        "username": "Lina George",
        "acctype": 0,
        "validuntil": 8146061670
    },
    {
        "uid": "13c2627d",
        "pincode": "1234",
        "username": "Fields Macdonald",
        "acctype": 0,
        "validuntil": 8564191761
    },
    {
        "uid": "61f7214b",
        "pincode": "1234",
        "username": "Mollie Dean",
        "acctype": 1,
        "validuntil": 3201049552
    },
    {
        "uid": "94f4198b",
        "pincode": "1234",
        "username": "Silva Lawrence",
        "acctype": 0,
        "validuntil": 3967606320
    },
    {
        "uid": "52c4392d",
        "pincode": "1234",
        "username": "Tyler Gonzalez",
        "acctype": 0,
        "validuntil": 7258523602
    },
    {
        "uid": "74f9928c",
        "pincode": "1234",
        "username": "Summers Whitehead",
        "acctype": 0,
        "validuntil": 3326998231
    },
    {
        "uid": "52f6266c",
        "pincode": "1234",
        "username": "Misty Williams",
        "acctype": 1,
        "validuntil": 1760784347
    },
    {
        "uid": "71f4745c",
        "pincode": "1234",
        "username": "Sanchez Osborn",
        "acctype": 1,
        "validuntil": 9210015456
    },
    {
        "uid": "65a5218c",
        "pincode": "1234",
        "username": "Brown Glover",
        "acctype": 0,
        "validuntil": 6616351487
    },
    {
        "uid": "82f6569d",
        "pincode": "1234",
        "username": "Pat Flowers",
        "acctype": 1,
        "validuntil": 2176624484
    },
    {
        "uid": "21a6381c",
        "pincode": "1234",
        "username": "Tanya Mcmillan",
        "acctype": 1,
        "validuntil": 7135500634
    },
    {
        "uid": "62c8616b",
        "pincode": "1234",
        "username": "Norma Perry",
        "acctype": 0,
        "validuntil": 8382065121
    },
    {
        "uid": "23f2325b",
        "pincode": "1234",
        "username": "Etta Mooney",
        "acctype": 1,
        "validuntil": 5466868107
    },
    {
        "uid": "33a5116b",
        "pincode": "1234",
        "username": "Marissa Waller",
        "acctype": 0,
        "validuntil": 8998269761
    },
    {
        "uid": "23c7176b",
        "pincode": "1234",
        "username": "Kari Gutierrez",
        "acctype": 1,
        "validuntil": 3100510208
    }
]

var configfile = {
    "command": "configfile",
    "network": {
        "bssid": "aa:bb:Cc:dd:ee",
        "ssid": "SMC",
        "wmode": "0",
        "pswd": "33355555",
        "offtime": "180"
    },
    "hardware": {
        "readertype": "0",
        "wgd0pin": "4",
        "wgd1pin": "5",
        "sspin": "15",
        "rfidgain": "32",
        "rtype": "1",
        "rpin": "16",
        "rtime": "300"
    },
    "general": {
        "hostnm": "esp-rfid",
        "restart": "86400",
        "pswd": "admin",
        "version": "v0.6",
        "openinghours": [
            "111111111111111111111111",
            "000000000000011100000000",
            "111111111111111111111111",
            "000011111000000000000000",
            "000000000000000000000000",
            "000000000111111000000000",
            "000000000000000000000000",
        ]
    },
    "mqtt": {
        "host": "",
        "port": "",
        "topic": "",
        "user": "",
        "pswd": ""
    },
    "ntp": {
        "server": "pool.ntp.org",
        "interval": "30",
        "timezone": "0"
    }
};

function remove(uidKey) {
    for (var i = 0; i < users.length; i++) {
        if (users[i].uid === uidKey) {
            console.log("[ INFO ] Removed: " + JSON.stringify(users[i]));
            users.splice(i, 1);
        }
    }
}

function updateuser(obj) {
    for (var i = 0; i < users.length; i++) {
        if (users[i].uid === obj.uid) {
            console.log("[ INFO ] Old User settings: " + JSON.stringify(users[i]));
            users.splice(i, 1);
            break;
        }
    }
    var newdata = {};
    newdata.uid = obj.uid;
    newdata.pincode = obj.pincode;
    newdata.username = obj.user;
    newdata.acctype = obj.acctype;
    newdata.validuntil = obj.validuntil;
    console.log("[ INFO ] New User settings: " + JSON.stringify(newdata));
    users.push(newdata);
    var res = {
        "command": "result",
        "resultof": "userfile",
        "result": true
    };
    wss.broadcast(res);
}


function sendUserList(page) {
    var datatosend = {};
    datatosend.command = "userlist"
    datatosend.page = page;
    datatosend.haspages = Math.ceil(users.length / 15);
    datatosend.list = [];
    var zero = 0;
    for (var i = ((page - 1) * 15); i < (page * 15); i++) {
        if (typeof users[i] !== "undefined") {
            datatosend.list[zero++] = users[i];
        }
    }
    wss.broadcast(datatosend);
    var res = {
        "command": "result",
        "resultof": "userlist",
        "result": true
    };
    wss.broadcast(res);
}

function sendEventLog() {
    wss.broadcast(eventlog);
    var res = {
        "command": "result",
        "resultof": "eventlist",
        "result": true
    };
    wss.broadcast(res);
}

function sendStatus() {
    var stats = {
        "command": "status",
        "heap": 30000,
        "chipid": "emu413",
        "cpu": "80/160",
        "availsize": 555555,
        "availspiffs": 445555,
        "spiffssize": 888888,
        "uptime": "1 Day 6 Hours",
        "ssid": "emuSSID",
        "dns": "8.8.8.8",
        "mac": "EM:44:11:33:22",
        "ip": "192.168.2.2",
        "gateway": "192.168.2.1",
        "netmask": "255.255.255.0"
    };
    wss.broadcast(stats);
}

wss.on('connection', function connection(ws) {
    ws.on("error", () => console.log("[ WARN ] WebSocket Error - Assume a client is disconnected."));
    ws.on('message', function incoming(message) {
        var obj = JSON.parse(message);
        console.log("[ INFO ] Got Command: " + obj.command);
        switch (obj.command) {
            case "remove":
                console.log("[ INFO ] Removing " + obj.uid);
                remove(obj.uid);
                break;
            case "configfile":
                configfile = obj;
                console.log("[ INFO ] New configuration file is recieved");
                configfile = obj;
                break;
            case "userlist":
                console.log("[ INFO ] Sending User List, page: " + obj.page);
                sendUserList(obj.page);
                break;
            case "status":
                console.log("[ INFO ] Sending Fake Emulator Status");
                sendStatus();
                break;
            case "userfile":
                console.log("[ INFO ] User Update " + obj.uid);
                updateuser(obj);
                break;
            case "testrelay":
                console.log("[ INFO ] Test relay button");
                process.stderr.write("\007");
                break;
            case "getlatestlog":
                console.log("[ INFO ] Sending latest log file");
                wss.broadcast(latestlog);
                var res = {
                    "command": "result",
                    "resultof": "latestlist",
                    "result": true
                };
                wss.broadcast(res);
                break;
            case "scan":
                console.log("[ INFO ] Sending Fake Wireless Networks");
                wss.broadcast(networks);
                break;
            case "gettime":
                console.log("[ INFO ] Sending time");
                var res = {};
                res.command = "gettime";
                res.epoch = Math.floor((new Date).getTime() / 1000);
                res.timezone = configfile.timezone;
                wss.broadcast(res);
                break;
            case "settime":
                console.log("[ INFO ] Setting time (fake)");
                var res = {};
                res.command = "gettime";
                res.epoch = Math.floor((new Date).getTime() / 1000);
                res.timezone = configfile.timezone;
                wss.broadcast(res);
                break;
            case "getconf":
                console.log("[ INFO ] Sending configuration file (if set any)");
                wss.broadcast(configfile);
                break;
            case "geteventlog":
                console.log("[ INFO ] Sending eventlog");
                sendEventLog();
                break;
            default:
                console.log("[ WARN ] Unknown command ");
                break;
        }
    });
});
